packages/gatsby-plugin-image/src/components/gatsby-image.browser.tsx

```diff
@@ -0,0 +1,187 @@
+/* eslint-disable no-unused-expressions */
+import React, {
+  ElementType,
+  useEffect,
+  useRef,
+  FunctionComponent,
+  ImgHTMLAttributes,
+  useState,
+  RefObject,
+} from "react"
+import {
+  getWrapperProps,
+  hasNativeLazyLoadSupport,
+  storeImageloaded,
+} from "./hooks"
+import { ILayoutWrapperProps } from "./layout-wrapper"
+import { PlaceholderProps } from "./placeholder"
+import { MainImageProps } from "./main-image"
+
+export type GatsbyImageProps = Omit<
+  ImgHTMLAttributes<HTMLImageElement>,
+  "placeholder" | "onLoad"
+> & {
+  alt: string
+  as?: ElementType
+  layout: ILayoutWrapperProps["layout"]
+  className?: string
+  height: number
+  images: Pick<MainImageProps, "sources" | "fallback">
+  placeholder: Pick<PlaceholderProps, "sources" | "fallback">
+  width: number
+  onLoad?: () => void
+  onError?: () => void
+  onStartLoad?: Function
+}
+
+let shownWarning = false
+
+export const GatsbyImageHydrator: FunctionComponent<GatsbyImageProps> = function GatsbyImageHydrator({
+  as: Type = `div`,
+  style,
+  className,
+  layout = `fixed`,
+  width,
+  height,
+  images,
+  onStartLoad,
+  onLoad: customOnLoad,
+  ...props
+}) {
+  const root = useRef<HTMLElement>()
+  const hydrated = useRef(false)
+  const unobserveRef = useRef<
+    ((element: RefObject<HTMLElement | undefined>) => void) | null
+  >(null)
+  const lazyHydrator = useRef<(() => void) | null>(null)
+  const ref = useRef<HTMLImageElement | undefined>()
+  const [isLoading, toggleIsLoading] = useState(hasNativeLazyLoadSupport)
+  const [isLoaded, toggleIsLoaded] = useState(false)
+
+  if (!global.GATSBY___IMAGE && !shownWarning) {
```

Comment:
Should this be `hasShownWarning`? My read is that it's keeping tracking of whether or not the warning has already appeared?

Reply:
It seems pretty clear now, though I guess adding `has` is slightly clearer